kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -0,0 +1,437 @@
+
+    val resolvedTableAdapter = {
+      if (tableVersionOpt.isPresent) {
+        tableManager.getResolvedTableAdapterAtVersion(
+          customEngine,
+          tablePath,
+          tableVersionOpt.get())
+      } else {
+        tableManager.getResolvedTableAdapterAtLatest(customEngine, tablePath)
+      }
+    }
+    resolvedTableAdapter.getScanBuilder().buildPaginated(pageSize, pageTokenOpt)
+  }
+
+  private def collectPaginatedBatches(
+      paginatedIter: PaginatedScanFilesIterator): Seq[FilteredColumnarBatch] = {
+    val buffer = collection.mutable.Buffer[FilteredColumnarBatch]()
+    while (paginatedIter.hasNext) {
+      val batch = paginatedIter.next()
+      buffer += batch
+    }
+    buffer
+  }
+
+  private def validatePageResults(
+      batches: Seq[FilteredColumnarBatch],
+      expectedFileCount: Int,
+      expectedBatchCount: Int,
+      testName: String): Unit = {
+    assert(batches.nonEmpty)
+    val fileCounts: Seq[Long] = batches.map(_.getPreComputedNumSelectedRows.get().toLong)
+    val totalFileCountsReturned = fileCounts.sum
+
+    assert(fileCounts.length == expectedBatchCount)
+    assert(totalFileCountsReturned == expectedFileCount)
+
+    logger.info(s"$testName: Total num batches returned in page one = ${fileCounts.length}")
+    logger.info(s"$testName: Total num Parquet Files fetched in page one = " +
+      s"$totalFileCountsReturned")
+  }
+
+  private def validatePageToken(
+      paginatedIter: PaginatedScanFilesIterator,
+      expectedLogFilePath: String,
+      expectedRowIndex: Long,
+      testName: String): Unit = {
+    val lastReadLogFilePath = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReadLogFilePath
+    val lastReturnedRowIndex = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReturnedRowIndex
+
+    assert(lastReadLogFilePath.endsWith(expectedLogFilePath))
+    assert(lastReturnedRowIndex == expectedRowIndex)
+
+    logger.info(s"$testName: New PageToken: lastReadLogFileName = $lastReadLogFilePath")
+    logger.info(s"$testName: New PageToken: lastReadRowIndex = $lastReturnedRowIndex")
+  }
+
+  private def runSinglePaginationTestCase(
+      testName: String,
+      tablePath: String,
+      tableVersionOpt: Optional[Long],
+      pageSize: Long,
+      expectedFileCount: Int,
+      expectedBatchCount: Int,
+      expectedLogFileName: String,
+      expectedRowIndex: Int): Unit = {
+    val paginatedScan = createPaginatedScan(
+      tablePath = tablePath,
+      tableVersionOpt = tableVersionOpt,
+      pageSize = pageSize)
+
+    val paginatedIter = paginatedScan.getScanFiles(customEngine)
+    val returnedBatchesInPage = collectPaginatedBatches(paginatedIter)
+
+    validatePageResults(returnedBatchesInPage, expectedFileCount, expectedBatchCount, testName)
+    validatePageToken(paginatedIter, expectedLogFileName, expectedRowIndex, testName)
+
+    paginatedIter.close()
+  }
+
+  // TODO: test call hasNext() twice
```

Comment:
btw, please make sure you track this TODO. have it in some project plan or some github issue so we do not forget it 👍